car-bridge/src/messaging.rs

```diff
@@ -0,0 +1,172 @@
+        )
+        .map_err_with("Failed to create MQTT client.")?;
+
+        // Get the stream before connecting the client, as otherwise messages
+        // may be lost. If the Car Bridge restarts while the number of
+        // outstanding messages is larger than the buffer size, messages may be
+        // lost.
+
+        let receiver = client.get_stream(MQTT_CLIENT_BUFFER_SIZE);
+
+        client
+            .connect(
+                ConnectOptionsBuilder::new()
+                    .mqtt_version(MQTT_VERSION_5)
+                    .automatic_reconnect(Duration::from_secs(1), Duration::from_secs(60))
+                    .finalize(),
+            )
+            .await
+            .map_err_with("Could not connect to MQTT broker.")?;
+
+        Ok(Self { client, receiver, is_subscribed: false })
+    }
+}
+
+#[async_trait]
+impl Subscriber for MqttMessaging {
+    type Message = Message;
+    type Topic = String;
+    type Error = Error;
+
+    /// Subscribes to a topic on the MQTT broker. Currently, this can only be
+    /// invoked once on `MqttMessaging`, due to the structure of the underlying
+    /// client.
+    async fn subscribe<'a>(
+        &'a mut self,
+        topic: String,
+    ) -> Result<BoxStream<'static, Self::Message>, Error> {
+        // TODO: By broadcasting the events on the underlying `Receiver` and
+        // filtering the broadcasted events by their topic name, we can support
+        // multiple subscriptions. Since this is currently not needed, we do not
+        // add said complexity to the implementation but fail at runtime
+        // instead.
+
+        if self.is_subscribed {
+            return Err(Error::new("Already receiving messages. It is currently not possible to subscribe multiple times."));
+        }
+
+        self.is_subscribed = true;
+
+        // C2D messages must be delivered with QOS 2, as we cannot assume that
+        // the fulfill requests they contain are always idempotent.
+
+        self.client
+            .subscribe(topic.clone(), QOS_2)
+            .await
+            .map_err_with("Could not subscribe to topic for receiving C2D messages.")?;
+
+        let mut receiver = self.receiver.clone();
+
+        let s = stream! {
+            while let Some(message) = receiver.next().await {
+                if let Some(message) = message {
+                    yield message;
+                }
+                else {
+                    // Automatic reconnect is used when connecting the
+                    // `AsyncClient`.
+                    info!("Connection temporarily lost. Attempting automatic reconnect.");
+                }
+            }
+        };
+
+        Ok(s.boxed())
+    }
+}
+
+#[async_trait]
+impl Publisher for MqttMessaging {
+    type Message = MessageBuilder;
+    type Topic = String;
+    type Error = Error;
+
+    /// Publish a message to an MQTT broker.
+    async fn publish(&self, topic: Self::Topic, message: Self::Message) -> Result<(), Error> {
```

Comment:
nit: Use `Self::` consistently:

```suggestion
    async fn publish(&self, topic: Self::Topic, message: Self::Message) -> Result<(), Self::Error> {
```